### PR TITLE
refactor(bitbucket-tags): move datasource to class

### PR DESCRIPTION
**Pull request by secustor**
## Changes:
Moves the BitBucketTags datasource to the class based approach

## Context:
#8647

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/datasource/go/__snapshots__/index.spec.ts.snap [hunk 1]
```diff
@@ -329,6 +329,7 @@ Array [
 
 exports[`datasource/go/index getReleases support bitbucket tags 1`] = `
 Object {
+  "registryUrl": "https://bitbucket.org",
```
Comment (viceice) [inline lib/datasource/go/__snapshots__/index.spec.ts.snap:332]: Why this change happened?

File: lib/datasource/bitbucket-tags/index.ts [hunk 2]
```diff
@@ -1,133 +1,133 @@
...
+import { cache } from '../../util/cache/package/decorator';
 import { BitbucketHttp } from '../../util/http/bitbucket';
 import { ensureTrailingSlash } from '../../util/url';
+import { Datasource } from '../datasource';
 import type { DigestConfig, GetReleasesConfig, ReleaseResult } from '../types';
 import { BitbucketCommit, BitbucketTag } from './types';
 
-const bitbucketHttp = new BitbucketHttp();
+export class BitBucketTagsDatasource extends Datasource {
+  bitbucketHttp = new BitbucketHttp();
```
Comment (viceice) [inline lib/datasource/bitbucket-tags/index.ts:10]: @JamieMagee I think we should make `Datasource` more generic to allow override `http` property. So we don't have two instances of `Http` class. 🤔

File: lib/datasource/go/__snapshots__/index.spec.ts.snap (hunk 1 above)
Comment (secustor) [reply lib/datasource/go/__snapshots__/index.spec.ts.snap:332]: I had to add this line so that no snapshots of the datasource changes. https://github.com/renovatebot/renovate/pull/10957/files#diff-74ff599c343f69422bac127d96420da26aeb58f0274040daedb002a1c81146f8R58

This change affects the Go datasource too as it uses bitbucket-tags. In this case only the default registry is returned explicitly now.

File: lib/datasource/bitbucket-tags/index.ts (hunk 2 above)
Comment (rarkins) [reply lib/datasource/bitbucket-tags/index.ts:10]: New issue?

Comment (JamieMagee) [reply lib/datasource/bitbucket-tags/index.ts:10]: Something [like this](https://www.typescriptlang.org/play?#code/IYIwzgLgTsDGEAJYBthjAgIsCaD2ArlLAKYA8AKgiQB4QkB2AJhgBIQQAOCAvAu1wB8AWABQAbzEJpCTlDz14JJkjwNIUAvDxQAFJwIhkAS1gIoJYEzXIAngmNMAXAg3GGAcwA0CCLc4kLroMJADuCLqOLm6eAJS8gggUsfGSojIZvgAWxmAAdFkc3Hwh4X4BkUyxANxSMgC+YnXShVwuFLWijaJiKGhsRQhpGQZGpqrq0FoQOvqGJmYWVjb2Ua7Q7h6x4t3dfegIAELGECBaANYkEALctPTMA1xD3b2oB8enF1fYuGCExCRqHRGCwsDh8ERSGQPmdYJdrkVEsMZM1ZPNxpAcOMltYGHYHCo+AByEAnWHwomdDKo2BqDTTWapVEZMAEAJ6GFfCA-CEAvKOHycuFXG6xVG7JqifYYACCDAUWRIUB5f0hgLuIIwKv+pCGqNGC3WWMWllx+McvAQROA8ogiqglNRNLpU20eiZ6Uy0lZ7N0coVSu1av5TB8ovFYnqQA)?

I'm not sure how to set a default value for the type parameter this way.

Comment (viceice) [reply lib/datasource/bitbucket-tags/index.ts:10]: Yeah. Can we make `Http` class the default value if not passed?

Comment (JamieMagee) [reply lib/datasource/bitbucket-tags/index.ts:10]: Created #10991 to summarise this discussion, and marking this thread as resolved for now.

File: lib/datasource/bitbucket-tags/index.ts [hunk 3]
```diff
@@ -1,133 +1,133 @@
...
-  newValue?: string
-): Promise<string | null> {
-  if (newValue?.length) {
-    return getTagCommit(registryUrl, repo, newValue);
+  // getReleases fetches list of tags for the repository
+  @cache({
+    namespace: BitBucketTagsDatasource.cacheNamespace,
+    key: ({ registryUrl, lookupName }: GetReleasesConfig) =>
+      BitBucketTagsDatasource.getCacheKey(registryUrl, lookupName, 'tags'),
+    ttlMinutes: 10,
```
Comment (JamieMagee) [inline lib/datasource/bitbucket-tags/index.ts:42]: What do you think about extending this to the default of 30 minutes?

File: lib/datasource/bitbucket-tags/index.ts [hunk 4]
```diff
@@ -1,133 +1,133 @@
...
-    version: name,
-    gitRef: name,
-    releaseTimestamp: target?.date,
-  }));
-
-  const cacheMinutes = 10;
-  await packageCache.set(cacheNamespace, cacheKey, dependency, cacheMinutes);
-  return dependency;
-}
+  static readonly cacheNamespace = 'datasource-bitbucket';
```
Comment (JamieMagee) [inline lib/datasource/bitbucket-tags/index.ts:20]:
```suggestion
  static readonly cacheNamespace = `datasource-${BitBucketTagsDatasource.id}`;
```

File: lib/datasource/bitbucket-tags/index.ts (hunk 3 above)
Comment (JamieMagee) [reply lib/datasource/bitbucket-tags/index.ts:42]: Same for all the TTLs below as well.

**Review by JamieMagee — DISMISSED** [review]
Couple of small comments. Overall good to go!

File: lib/datasource/bitbucket-tags/index.ts (hunk 3 above)
Comment (secustor) [reply lib/datasource/bitbucket-tags/index.ts]: I see no reason not to. At the moment they are simply mimicking the previous logic.

**Review by JamieMagee — APPROVED** [review]

**Review by viceice — APPROVED** [review]
